Add instructions for targeting .NET Core with MSBuild projects

dsplaisted opened the pull request:
This updates our [outdated instructions](http://aka.ms/port2core) on how to target .NET Core with MSBuild projects.  As described in the instructions, most people today should just use project.json, but these instructions should be useful for people who are porting existing projects that use MSBuild or who have customized their build process in ways that project.json doesn't support well.

@bleroy @sayedihashimi @weshaggard @ericstj @mellinoe @davkean

mellinoe commented on Documentation/project-docs/target-dotnetcore-with-msbuild.md:
```diff
@@ -0,0 +1,119 @@
+
+- [Visual Studio 2015 Update 2 or higher](https://www.visualstudio.com/downloads/download-visual-studio-vs)
+- [.NET Core tools for Visual Studio](https://www.visualstudio.com/downloads/download-visual-studio-vs)
+- NuGet Visual Studio extension [v3.5.0-beta](https://dist.nuget.org/visualstudio-2015-vsix/v3.5.0-beta/NuGet.Tools.vsix) or later
```
I think all of this works with the nuget version you can get from the extensions gallery, in case folks don't want to install a beta plugin.

weshaggard replied:
I believe most of this will work with nuget 3.4 right?

joperezr replied:
I'm not 100% sure but I think that this is needed because netstandard1.5 is not recognized as a framework yet, but using the beta extension will fix this.

mellinoe replied:
Actually, its `netstandard1.6` that wont work without the beta extension, I think.

dsplaisted replied:
The NuGet 3.4 VS Extension worked for me until I tried adding references to packages which required adding the imports for `dnxcore50` and `portable-net452+win81` to the `netstandard1.5` framework.  Then if the restore was done in VS, it would succeed, but VS wouldn't pick up any references so you'd get errors that `System.Object` etc wasn't defined.

mellinoe commented on Documentation/project-docs/target-dotnetcore-with-msbuild.md:
```diff
@@ -0,0 +1,119 @@
+    - Change the Output type to "Console Application" in the Application tab
+    - In the "Build" tab, select "All Configurations" and change the "Platform Target" to "x64"
+- Add runtimes and dependencies on the runtime and host to `project.json`
+    - Add a `runtimes` section with the following runtimes: `win10-x64`, `ubuntu.14.04-x64`, and `osx.10.10-x64`
```
The last two wont be useful unless you also add some project configuration to control `BaseNuGetRuntimeIdentifier` (harcoded above).

dsplaisted replied:
> The last two wont be useful unless you also add some project configuration to control `BaseNuGetRuntimeIdentifier` (hardcoded above).

Yeah, I thought it was simpler just to leave them in instead of explaining when you did or didn't need them.

mellinoe commented on Documentation/project-docs/target-dotnetcore-with-msbuild.md:
```diff
@@ -0,0 +1,119 @@
+    - In the "Build" tab, select "All Configurations" and change the "Platform Target" to "x64"
+- Add runtimes and dependencies on the runtime and host to `project.json`
+    - Add a `runtimes` section with the following runtimes: `win10-x64`, `ubuntu.14.04-x64`, and `osx.10.10-x64`
+    - Add a dependency on `Microsoft.NETCore.Runtime` version `1.0.2-rc2-24027` and `Microsoft.NETCore.TestHost` version `1.0.0-rc2-24027`
```
We might just want to say `Microsoft.NETCore.Runtime.CoreCLR` here. We don't produce this "meta-package" anymore.

mjsabby commented on Documentation/project-docs/target-dotnetcore-with-msbuild.md:
```diff
@@ -0,0 +1,119 @@
+- Open the project's XML for editing (in Visual Studio, right click on the project -> Unload Project, right click again -> Edit MyProj.csproj)
+    - Remove the `ProjectTypeGuids` property
+    - Add the following property: `<BaseNuGetRuntimeIdentifier>win10</BaseNuGetRuntimeIdentifier>`
+    - Add the following at the end of the file (after the import of `Microsoft.Portable.CSharp.Targets`:
```
Leaving a comment saying this enables debugging using the Managed CoreCLR Debugging Engine will help people grok this seemingly voodoo guid and other params.

mjsabby commented:
👍 -- this is almost exactly Bing's setup, so I think this will truly be beneficial to lots of existing customers.

Few thoughts:

In creating a .net core console application, you're referring to Git repositories, can't we just put @mellinoe's props and targets in a NuGet package and have the user also specify that in addition to all the other stuff you're asking them to put in project.json?

Secondly, the coretemplate repo has the same corefx GetDirectoryNameOfFileAbove. I don't think most people do that, I think they'd understand better if you just said include this 1 props file and 1 targets file and things will light up. The dirs.targets and dirs.props requires extra thought on the developer which really isn't needed.

I really like this doc, primarily because it exposes the guts of how it all works, and the targets and props from NuGet I think would make it even easier to onboard.

Finally, will these instructions work on MSBuild XPlat? The portable Csharp targets are there right? If so, it'd be nice to add a note and link to that repo, and let's also call out the target file imports are case-sensitive file paths.

mellinoe commented:
> Finally, will these instructions work on MSBuild XPlat? The portable Csharp targets are there right? If so, it'd be nice to add a note and link to that repo, and let's also call out the target file imports are case-sensitive file paths.

Projects cloned from coretemplate work with x-plat msbuild, with one caveat: you need an assembly from buildtools for the [ResolveNuGetPackageAssets](https://github.com/dotnet/buildtools/blob/30e9bdae21d38805edec62be438ad5306374f770/src/Microsoft.DotNet.Build.Tasks/PrereleaseResolveNuGetPackageAssets.cs) task, imported [here in coretemplate](https://github.com/mellinoe/corebuild/blob/ac59741e0732b5387c06a348e804938f234c72e0/corebuild.targets#L34). This is not needed on Windows because NuGet installs some built-in tasks, including that one (at C:\Program Files (x86)\MSBuild\Microsoft\NuGet). Maybe there will be an official version of this task soon that works on .NET Core.

weshaggard commented:
cc @joperezr

weshaggard commented on Documentation/project-docs/target-dotnetcore-with-msbuild.md:
```diff
@@ -0,0 +1,119 @@
+We expect the first version of the .NET Core tools that use MSBuild to ship along with Visual Studio "15".  However, it is possible to use MSBuild for .NET Core
+projects today, and this page shows how.
+
+We recommend that most people targeting .NET Core today use the default tooling experience with project.json.  This is because we haven't yet added support to MSBuild
```
I would say we recommend new projects targeting .NET Core today to use the default tooling experience with project.json.

mellinoe replied:
Isn't that what this says, or did you mean something else?

> We recommend that most people targeting .NET Core today use the default tooling experience with project.json.

weshaggard replied:
The keyword for me is "new". I don't want to recommend existing customers with msbuild projects switch to project.json.

dsplaisted replied:
> The keyword for me is "new". I don't want to recommend existing customers with msbuild projects switch to project.json.

I've updated the wording.  Let me know if it looks good now.

weshaggard commented on Documentation/project-docs/target-dotnetcore-with-msbuild.md:
```diff
@@ -0,0 +1,119 @@
+Creating a .NET Core console application
+========================================
+Building a console application for .NET Core requires some customization of the MSBuild build process.  A good option to start with is
+[coretemplate](https://github.com/mellinoe/coretemplate).  This is a basic project which brings in MSBuild targets for .NET Core.  
```
I have updated the sample at https://github.com/dotnet/corefxlab/tree/master/samples/NetCoreSample, based on RC2 so it all works as well. I've not looked into @mellinoe sample deeply but the sample in corefxlab doesn't require any extra targets just some hacking in the project files themselves and give examples for out to cross-compile file x-plat.

weshaggard replied:
Also when I updated it for RC2 I switch it to using dotnet.exe as the host which is more closely aligned with the dotnet cli approach.

weshaggard commented on Documentation/project-docs/target-dotnetcore-with-msbuild.md:
````diff
@@ -0,0 +1,119 @@
+        ```json
+        {
+            "dependencies": {
+                "NETStandard.Library": "1.5.0-rc2-24027",
````
See my sample https://github.com/dotnet/corefxlab/blob/master/samples/NetCoreSample/CoreApp/project.json#L3 I think this should be switched to "Microsoft.NETCore.App" and use dotnet.exe now.

weshaggard commented on Documentation/project-docs/target-dotnetcore-with-msbuild.md:
```diff
@@ -0,0 +1,119 @@
+        {
+            "dependencies": {
+                "NETStandard.Library": "1.5.0-rc2-24027",
+                "Microsoft.NETCore.TestHost": "1.0.0-rc2-24027",
```
We should try and avoid people using TestHost package as it is only for testing and we aren't even going to be shipping it stable. dotnet.exe should be the host we point folks to.

weshaggard commented on Documentation/project-docs/target-dotnetcore-with-msbuild.md:
```diff
@@ -0,0 +1,119 @@
+            "dependencies": {
+                "NETStandard.Library": "1.5.0-rc2-24027",
+                "Microsoft.NETCore.TestHost": "1.0.0-rc2-24027",
+                "Microsoft.NETCore.Runtime": "1.0.2-rc2-24027"
```
I do think we should use Microsoft.NETCore.App but if you stick with the runtime specifically you should switch to "Microsoft.NETCore.Runtime.CoreCLR" as we are removing "Microsoft.NETCore.Runtime" package for RTM.

weshaggard commented:
@dsplaisted Thanks for spending some time updating the documents. 
From my comments you can see I would prefer we push people to the shipping meta-packages even for msbuild stuff so I do highly recommend we use "Microsoft.NETCore.App" and dotnet.exe as our host for this, as I did in my sample update.

dsplaisted commented:
I've updated the instructions for creating a .NET Core console application to start with a .NET Framework console app project instead of a portable library, and to use the `Microsoft.NETCore.App` package.  I've also adjusted the recommendation for using project.json/xproj to only apply to new projects, and added the CoreApp sample as another option to start off with.

weshaggard commented:
LGTM Thanks @dsplaisted.

weshaggard commented:
please squash and then merge.

mellinoe commented:
LGTM after squash.

dsplaisted commented:
@bleroy suggested putting this information in the .NET Core docs site, so I've created dotnet/core-docs#550 for that.  Once that goes in, I'll update this pull request to point to the documentation there.

joshfree commented:
@dsplaisted can you please squash these commits?

stephentoub commented:
I'm going to squash and merge this.  If you want to add a link, it can be done in a separate PR. Thanks.
